scripts/dndrks/less_concepts.lua

```diff
@@ -0,0 +1,803 @@
+clk = beatclock.new()
+clk_midi = midi.connect()
+clk_midi.event = clk.process_midi
+
+engine.name = "Passersby"
+passersby = require "mark_eats/passersby"
+
+-- this section is all maths + computational events
+
+-- maths: translate the seed integer to binary
+function seed_to_binary()
+  seed_as_binary = {}
+  for i = 0,7 do
+    table.insert(seed_as_binary, (seed & (2 ^ i)) >> i)
+  end
+end
+
+-- maths: translate the rule integer to binary
+function rule_to_binary()
+  rule_as_binary = {}
+  for i = 0,7 do
+    table.insert(rule_as_binary, (rule & (2 ^ i)) >> i)
+  end
+end
+
+-- maths: basic compare function, used in bang()
+function compare (s, n)
+  if type(s) == type(n) then
+        if type(s) == "table" then
+                  for loop=1, 3 do
+                    if compare (s[loop], n[loop]) == false then
+                        return false
+                    end
+                  end
+
+                return true
+        else
+            return s == n
+        end
+    end
+    return false
+end
+
+-- maths: scale seeds to the note pool + range selected
+function scale(lo, hi, received)
+  scaled = math.floor(((((received-1) / (256-1)) * (hi - lo) + lo)))
+end
+
+-- pack the seeds into clusters, compare these against neighborhoods to determine gates in iterate()
+function bang()
+  redraw()
+  seed_to_binary()
+  rule_to_binary()
+  seed_pack1 = {seed_as_binary[1], seed_as_binary[8], seed_as_binary[7]}
+  seed_pack2 = {seed_as_binary[8], seed_as_binary[7], seed_as_binary[6]}
+  seed_pack3 = {seed_as_binary[7], seed_as_binary[6], seed_as_binary[5]}
+  seed_pack4 = {seed_as_binary[6], seed_as_binary[5], seed_as_binary[4]}
+  seed_pack5 = {seed_as_binary[5], seed_as_binary[4], seed_as_binary[3]}
+  seed_pack6 = {seed_as_binary[4], seed_as_binary[3], seed_as_binary[2]}
+  seed_pack7 = {seed_as_binary[3], seed_as_binary[2], seed_as_binary[1]}
+  seed_pack8 = {seed_as_binary[2], seed_as_binary[1], seed_as_binary[8]}
+
+neighborhoods1 = {1,1,1}
+neighborhoods2 = {1,1,0}
+neighborhoods3 = {1,0,1}
+neighborhoods4 = {1,0,0}
+neighborhoods5 = {0,1,1}
+neighborhoods6 = {0,1,0}
+neighborhoods7 = {0,0,1}
+neighborhoods8 = {0,0,0}
+
+function com1()
+if compare (seed_pack1,neighborhoods1) then
+    out1 = (rule_as_binary[8] << 7) & 128
+  elseif compare (seed_pack1, neighborhoods2) then
+    out1 = (rule_as_binary[7] << 7) & 128
+  elseif compare (seed_pack1, neighborhoods3) then
+    out1 = (rule_as_binary[6] << 7) & 128
+  elseif compare (seed_pack1, neighborhoods4) then
+    out1 = (rule_as_binary[5] << 7) & 128
+  elseif compare (seed_pack1, neighborhoods5) then
+    out1 = (rule_as_binary[4] << 7) & 128
+  elseif compare (seed_pack1, neighborhoods6) then
+    out1 = (rule_as_binary[3] << 7) & 128
+  elseif compare (seed_pack1, neighborhoods7) then
+    out1 = (rule_as_binary[2] << 7) & 128
+  elseif compare (seed_pack1, neighborhoods8) then
+    out1 = (rule_as_binary[1] << 7) & 128
+  else out1 = (0 << 7) & 128
+  end
+end
```

Comment:
looks like indentation is wrong in this function and below?

Reply:
resolved in update!